### ISSUE TITLE
Cast ID arguments to int32 using to_np_int32

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/tskit#192]
Any place we are providing a list of IDs as input to a function we should use the ``to_np_int32`` function defined in #185 to safely convert them.

These will be in the Tables API (node IDs, site IDs, etc), and where we compute stats (GNN, eg, takes lists of samples). 

Note there are still a number of functions that don't use the numpy C API (e.g., Tree.pairwise_diversity and TreeSequence.trees) but Python lists instead, which are overdue being updated (#64).

**Comment by hyanwong** [comment]
I think we should generalise the casting function, so that it works for e.g. `np.uint8` too (I've just hit this when inputting genotypes: [-1, 0] was invisibly converted to [255, 0], which was hard to track down). The alteration below seems a trivial adjustment of  https://github.com/tskit-dev/tskit/blob/57ff0abe4173611a396e9715632d544195880a75/python/tskit/tables.py#L1822 - but I think the function should be taken out of tables.py and put somewhere more general. Any suggestions as to where?

```
def safe_np_int_conversion(int_array, dtype=np.int32):
    """
    A few functions require arrays of type np.int32. To allow passing standard numpy
    integer arrays (dtype=np.int64) we cast but check bounds to avoid wrap-around
    conversion errors (numpy doesn't seem to provide this functionality)
    """
    int_array = np.array(int_array)
    if int_array.size == 0:
        return int_array.astype(dtype)  # Allow empty arrays of any type
    try:
        return int_array.astype(dtype, casting='safe')
    except TypeError:
        bounds = np.iinfo(dtype)
        if np.all(int_array >= bounds.min) and np.all(int_array <= bounds.max):
            # Raises a TypeError when we try to convert from, e.g., a float.
            return int_array.astype(dtype, casting='same_kind')
        else:
            raise OverflowError("Cannot convert safely to {} type".format(dtype))
```

**Comment by petrelharp** [comment]
Perhaps in a (new) `util.py`?

**Comment by hyanwong** [comment]
I will do this. Assign it to me.

**Comment by jeromekelleher** [comment]
Thanks! Done/.

**Comment by jeromekelleher** [comment]
Does #245 cover *every* usage of numpy int arrays as input @hyanwong?

**Comment by hyanwong** [comment]
As you may have seen, I have only written tests for [np.int32, np.uint32, np.int8, np.uint8](https://github.com/tskit-dev/tskit/blob/d6191f981573d2322c991ef50b9a96db72419232/python/tests/test_util.py#L36). I should probably expand this, which is easy to do.

I *think* I have the generic logic correct, but it's not tested for *every* type.

**Comment by jeromekelleher** [comment]
The function itself is good, no need to test it for other dtypes. Are we using it to convert all inputs into tskit (i.e., every tables API function, every stats input) is the question?

**Comment by jeromekelleher** [comment]
> The function itself is good, no need to test it for other dtypes. Are we using it to convert all inputs into tskit (i.e., every tables API function, every stats input) is the question?

Where are we with this @hyanwong - are we filtering all integer array inputs through this function?

**Comment by hyanwong** [comment]
No, I haven't changed all the APIs to use this yet. Sorry - slipped off my radar. Is there a simple way to find all those that need converting?

**Comment by hyanwong** [comment]
I'm guessing they all have test cases, so I could add the tests first (providing arrays that I expect to be convertible and those that I expect to be non-convertible), and then sprinkle in the conversions until the tests pass.

**Comment by jeromekelleher** [comment]
> No, I haven't changed all the APIs to use this yet. Sorry - slipped off my radar. Is there a simple way to find all those methods that need converting?

Not really - there's some leg work to be done going through the API (tables and stats APIs are the main user of numpy APIs.)

**Comment by jeromekelleher** [comment]
Closed in #245
